### PR TITLE
To Install helmcharts through Samples operator

### PR DESCRIPTION
**Pull request by fbm3307**
This PR is to update the Cluster Samples Operator CRD to have the Helmchart List . This is Part of Samples operator Downsizing Phase-2, where the S2i Sample helmcharts can be installed via CSO.
Refernce -[SO-112](https://issues.redhat.com/browse/SO-112)

### CODE REVIEW

**Comment by fbm3307** [comment]
closed the previous PR as it was very Old PR which i tried to revive and ran into problems.

File: samples/v1/types_config.go
Comment (bparees) [inline samples/v1/types_config.go:81]: is this supposed to convey that the default NS is openshift?  If so, please state it as such.

File: samples/v1/types_config.go
Comment (bparees) [inline samples/v1/types_config.go:68]: are users going to be able to edit this list?  What happens if a non-s2i chart is put in the list?

what happens when an item is deleted from this list? is the chart uninstalled from the cluster?

in general how is this list managed?

File: samples/v1/types_config.go
Comment (bparees) [inline samples/v1/types_config.go:83]: these validation rules don't seem aligned w/ kube namespace validation rules

File: samples/v1/types_config.go
Comment (bparees) [inline samples/v1/types_config.go:89]: are these validations rules based on helmchart releasename rules?

File: samples/v1/types_config.go
Comment (fbm3307) [reply samples/v1/types_config.go]: to make it easier and avoid confusion, i have updated the list of parameters and removed the namespace altogether, as it will be "openshift" only

File: samples/v1/types_config.go
Comment (fbm3307) [reply samples/v1/types_config.go:83]: I to make it easier and avoid confusion, i have updated the list of parameters and removed the namespace, as it will be "openshift" only

File: samples/v1/types_config.go
Comment (fbm3307) [reply samples/v1/types_config.go:91]: let me double check these rules

File: samples/v1/types_config.go
Comment (fbm3307) [reply samples/v1/types_config.go:68]: Users will be able to edit the list, if an item is deleted, chart wont be unistalled, but we wont manage it (same as current samples) 
We will have validation in place where we will only allow S2i helmcharts to be installed/managed by Samples and wont install NonS2i's

File: samples/v1/types_config.go
Comment (fbm3307) [reply samples/v1/types_config.go:93]: these are same as helmchart releasename

File: samples/v1/types_config.go
Comment (bparees) [reply samples/v1/types_config.go:68]: how are helmcharts identified as "s2i" helm charts?

what will happen if they specify a non-s2i helm chart? (will there be some status message/condition that makes it clear to the user that the chart is not going to be installed?)

whose permissions are used to install the chart?  (can a user leverage this api to elevate their own permissions by using the samples operator to instantiate a helmchart the user cannot instantiate themselves?)

Comment (fbm3307) [reply samples/v1/types_config.go:68]: The imagestreams has a tag as s2i, and we have identified the 12 which are required by the ODC team for day 1 user experience, so that 12 corresponding helmcharts only will allowed.
If they specifiy a Non-S2i, we plan on displaying a message that it won't be installed

Comment (bparees) [reply samples/v1/types_config.go:68]: > we plan on displaying a message that it won't be installed

displaying a message where?

File: samples/v1/types_config.go
Comment (fbm3307) [reply samples/v1/types_config.go]: ok updated it

File: samples/v1/types_config.go
Comment (fbm3307) [reply samples/v1/types_config.go:68]: > whose permissions are used to install the chart? (can a user leverage this api to elevate their own permissions by using the samples operator to instantiate a helmchart the user cannot instantiate themselves?)

it will use same permission used today to install the manifests for each samples. Since we are limiting the list of helm charts to install to only s2i charts then user won't be able to install any other charts through samples. Also the sample charts will be coming from Red Hat and will be able to verify digest and signature if one provided.

> displaying a message where?

The message will be displayed on CLI.

Comment (bparees) [reply samples/v1/types_config.go:68]: > it will use same permission used today to install the manifests for each samples. Since we are limiting the list of helm charts to install to only s2i charts then user won't be able to install any other charts through samples. Also the sample charts will be coming from Red Hat and will be able to verify digest and signature if one provided.

ok so it's a specific allow list of charts that can be installed via this api.

> The message will be displayed in logs of the controller

that's going to be pretty difficult for an admin to understand/debug.  Status information on the resource would be a better UX, listing the charts that could not be installed and why not (you can also use this for reporting other failures that might occur while installing a chart)

Comment (fbm3307) [reply samples/v1/types_config.go:68]: > that's going to be pretty difficult for an admin to understand/debug. Status information on the resource would be a better UX, listing the charts that could not be installed and why not (you can also use this for reporting other failures that might occur while installing a chart)
Thanks for raising the thoughtful questions, we discussed this and are thinking that rather than having the updatable helmchart list to install, how about we have a hardcoded list to install the s2i helmcharts same as it is for current samples, but give the customers the freedom to skip it[skiped helmchart list] as they can do it for current samples. this will avoid all the extra validation/confusion . the customer can skip the helmchart and install a newer version/chart or however they prefer by telling the CSO to not manage it

Comment (bparees) [reply samples/v1/types_config.go:68]: that sounds like a better/less error prone approach to me

**Review by dperaza4dustbit — APPROVED** [review]
/lgtm

**Comment by fbm3307** [comment]
@bparees can you please review this again? i have made the changes

**Review by bparees — COMMENTED** [review]
one nit on validation rules, otherwise lgtm as it follows the existing api pattern.

you'll need @openshift/api-reviewers to give the final approval once you resolve the validation (max items) discussion.

**Comment by bparees** [comment]
>      I Kept the validation as 12 max items because there are 12 S2i currently which we handle

yeah i assumed that was your logic, but it means you'll need to update this limit if you ever add more, and i don't really see that the having the validation/constraint in place is worth that cost.

**Comment by fbm3307** [comment]
> > ```
> >  I Kept the validation as 12 max items because there are 12 S2i currently which we handle
> > ```
> 
> yeah i assumed that was your logic, but it means you'll need to update this limit if you ever add more, and i don't really see that the having the validation/constraint in place is worth that cost.

i had the thought.. and i agree,  i shall remove the validation.

**Comment by fbm3307** [comment]
> > ```
> >  I Kept the validation as 12 max items because there are 12 S2i currently which we handle
> > ```
> 
> yeah i assumed that was your logic, but it means you'll need to update this limit if you ever add more, and i don't really see that the having the validation/constraint in place is worth that cost.

i have removed the validation in this

File: samples/v1/types_config.go
Comment (bparees) [inline samples/v1/types_config.go:74]: why is this `rep` and also why is it not `opt`?

Comment (fbm3307) [reply samples/v1/types_config.go]: Initially since the helmchart were to be installed by choice, i kept it as rep, as my understanding was i sholdn't keep it as optional since we need to install it.. but then i missed to change it back to opt after the discussion.. i have changed it now.

**Comment by bparees** [comment]
/lgtm

@openshift/api-reviewers for approval

**Comment by bparees** [comment]
removing my approval.  I only intended to lgtm, not approve.

/hold cancel

**Review by JoelSpeed — COMMENTED** [review]
What is the target release for this API?

File: samples/v1/types_config.go
Comment (fbm3307) [reply samples/v1/types_config.go:74]: I had initially kept a limit of 12, as the new changes would require operator to install 12 helmcharts.. but not sure about the future if we need to increase it and hence got a suggestion to remove the validation of 12 

> > ```
> >  I Kept the validation as 12 max items because there are 12 S2i currently which we handle
> > ```
> 
> yeah i assumed that was your logic, but it means you'll need to update this limit if you ever add more, and i don't really see that the having the validation/constraint in place is worth that cost.

Comment (JoelSpeed) [reply samples/v1/types_config.go:74]: How about a limit of 32? That's small enough that it stops someone going crazy with the field, but large enough that there's overhead from existing know use cases.

Or if you're confident, we could go for 16. Loosening the validation later is in theory easy assuming we control the consumers, which I think is just the samples operator right?

**Comment by JoelSpeed** [comment]
> targeting before ocp4.17 so mostly 4.16 want to implement the feature

Ok, I'd like to not merge the API change until 4.16 opens then. How large is the feature, if we merge this at the beginning of 4.16, are you confident the feature will be ready within the 4.16 time frame? Just wondering whether this merits putting behind a feature gate to ease the development flow if there are risks

File: samples/v1/types_config.go
Comment (fbm3307) [reply samples/v1/types_config.go:74]: Its just Samples Operator, and we are downsizing, so mostly 16 should suffice in my opinion. i will update it with max allowed as 16 .

**Comment by JoelSpeed** [comment]
/approve cancel
Lets come back to this once the 4.16 branch opens

**Comment by fbm3307** [comment]
@JoelSpeed can i get a review on this..4.16 is open now. and i have a Related [PR](https://github.com/openshift/cluster-samples-operator/pull/529) in review for this feature to be implemented

File: samples/v1/types_config.go
Comment (JoelSpeed) [inline samples/v1/types_config.go:68]: Maybe `managed` rather than `created/updated`?

File: samples/v1/types_config.go
Comment (JoelSpeed) [inline samples/v1/types_config.go:72]: This may want to be reworded to be more direct. It's currently a little confusing and if I were to stop reading half way through, I may not realise that I as the user, have to delete the content myself.

File: samples/v1/types_config.go
Comment (JoelSpeed) [inline samples/v1/types_config.go:75]: Is `Skipped` the best term here?

**Review by JoelSpeed — COMMENTED** [review]
If I were a user, how would I know what the helm chart names, that are valid to be skipped, are? Where would I be able to discover values I could use with this new API?

File: samples/v1/types_config.go
Comment (fbm3307) [reply samples/v1/types_config.go:75]: it has to be kept inline with previous terms as skippedtemplates 
here there will be by default installtions given so for customer to tell the operatorr to skip managing it used the term skipped hence skipped is okay to be used

Comment (JoelSpeed) [reply samples/v1/types_config.go:75]: > it has to be kept inline with previous terms as skippedtemplates

I would like to challenge that, why does it have to be kept inline? We are adding a new API, if there's a better name we could use, why not use it? We shouldn't propagate bad (I appreciate this is subjective) descriptions if we can think of better ones

Comment (bparees) [reply samples/v1/types_config.go:75]: `Unmanaged` might be a better term but sticking w/ the existing prefix is has its own intrinsic value both for users who have already learned/understood the behavior of the other fields that share the prefix, as well as for new users trying to understand the set of fields as whole and how they relate.  This is my number one reservation about the "we don't double down on mistakes" philosophy.  Change+inconsistency have their own costs.

To use a new term for the new field should require not just that the new name be "better", but that it is sufficiently better to justify creating the inconsistency/confusion.  (It would also help if we had evidence the existing name has been a source of pain/confusion.  Personally i'm of the school of thought that no one uses an api field w/o reading the api docs for the field at which point the field name, as long as it's not explicitly wrong/misleading and as long as the api doc is decent, doesn't matter a whole lot)

Comment (JoelSpeed) [reply samples/v1/types_config.go:75]: Definitely a fair point and I'm not saying Skipped is definitely a no go, I was kind of curious if we had any feedback, or perhaps, "I wish I'd called it X" moments. Opening up the discussion about potential other names. It feels to me like we should probably have some discussion about the name, even if at the end we come back to, no we had the best thing to begin with for XYZ reason, at least then we can justify the choice

File: samples/v1/types_config.go
Comment (fbm3307) [reply samples/v1/types_config.go]: updated it

**Comment by fbm3307** [comment]
> If I were a user, how would I know what the helm chart names, that are valid to be skipped, are? Where would I be able to discover values I could use with this new API?

Those are already mentioned in the customer doc

**Comment by fbm3307** [comment]
I have updated it now can you PTAL @JoelSpeed ?

**Comment by fbm3307** [comment]
ping @JoelSpeed

File: samples/v1/types_config.go
Comment (JoelSpeed) [inline samples/v1/types_config.go:75]: Where would I find the names of the helm charts?

Is there a format for the names of the helm charts? I assume they must fit the Kube standard [subdomain name](https://kubernetes.io/docs/concepts/overview/working-with-objects/names/#dns-subdomain-names) right?

Comment (fbm3307) [reply samples/v1/types_config.go:75]: The list of helmcharts is still being processed which are going to be available by default. and then the customer can use it skippedHelmcharts to skip them .The maximum would be 12, but they would come as the the imagestream owners convert them to helmcharts(by ocp4.17). once that is available we plan on making that list and put it to the product release notes for samples operator available to the customer. 
Also the customer would be able to see the list in openshift shift console in the  samples . 
This is as a part of Downsizing effort of CLuster Samples Operator and moving away from imagestreams.

Comment (JoelSpeed) [reply samples/v1/types_config.go:75]: It would be good to capture some of this context within the godoc. So for example, "To find available helm charts, check the openshift console ..." would be helpful context for users

**Comment by fbm3307** [comment]
closing this in favour of https://github.com/openshift/api/pull/1754

File: samples/v1/types_config.go
Comment (fbm3307) [reply samples/v1/types_config.go:80]: okay updated it in the new PR

**Pull request closed without merging**
